fggs/semirings.py

```diff
@@ -2,7 +2,8 @@
 from math import inf
 import torch_semiring_einsum, torch_semiring_einsum.utils
 from abc import ABC, abstractmethod
-from typing import Union
+from typing import Union, TYPE_CHECKING
+if TYPE_CHECKING: from fggs.indices import EmbeddedTensor
```

Comment:
Is this to avoid a circular import? Maybe it's a sign that something needs to be refactored somehow.